From 0716ca9e2a68a78a2af33bbadecd444b4d953189 Mon Sep 17 00:00:00 2001
From: Paul Harrison <pharrison@mythtv.org>
Date: Tue, 25 Mar 2014 17:12:10 +0000
Subject: [PATCH 2/2] MythMusic: add a 'Reset Database' button on the general
 settings page

This will clear the music tracks database tables so you can start afresh.
It doesn't clear the smart playlists or radio stream tables.
---
 .../mythmusic/mythmusic/generalsettings.cpp        |   53 +++++++++++++++++++-
 mythplugins/mythmusic/mythmusic/generalsettings.h  |    3 ++
 .../theme/default-wide/musicsettings-ui.xml        |    5 ++
 .../mythmusic/theme/default/musicsettings-ui.xml   |    7 ++-
 4 files changed, 66 insertions(+), 2 deletions(-)

diff --git a/mythplugins/mythmusic/mythmusic/generalsettings.cpp b/mythplugins/mythmusic/mythmusic/generalsettings.cpp
index 64efd49..82a473b 100644
--- a/mythplugins/mythmusic/mythmusic/generalsettings.cpp
+++ b/mythplugins/mythmusic/mythmusic/generalsettings.cpp
@@ -4,15 +4,19 @@
 
 // MythTV
 #include <mythcorecontext.h>
+#include <mythdialogbox.h>
 
+// MythMusic
+#include "musicplayer.h"
 #include "generalsettings.h"
+#include "musicdata.h"
 
 GeneralSettings::GeneralSettings(MythScreenStack *parent, const char *name)
         : MythScreenType(parent, name),
         m_musicAudioDevice(NULL),
         m_musicDefaultUpmix(NULL), m_musicCDDevice(NULL),
         m_nonID3FileNameFormat(NULL), m_ignoreID3Tags(NULL),
-        m_allowTagWriting(NULL), m_saveButton(NULL),
+        m_allowTagWriting(NULL), m_resetDBButton(NULL), m_saveButton(NULL),
         m_cancelButton(NULL)
 {
 }
@@ -36,6 +40,7 @@ bool GeneralSettings::Create()
     UIUtilE::Assign(this, m_nonID3FileNameFormat, "nonid3filenameformat", &err);
     UIUtilE::Assign(this, m_ignoreID3Tags, "ignoreid3tags", &err);
     UIUtilE::Assign(this, m_allowTagWriting, "allowtagwriting", &err);
+    UIUtilW::Assign(this, m_resetDBButton, "resetdatabase", &err);
     UIUtilE::Assign(this, m_saveButton, "save", &err);
     UIUtilE::Assign(this, m_cancelButton, "cancel", &err);
 
@@ -63,6 +68,9 @@ bool GeneralSettings::Create()
     if (allowTagWriting == 1)
         m_allowTagWriting->SetCheckState(MythUIStateType::Full);
 
+    if (m_resetDBButton)
+        connect(m_resetDBButton, SIGNAL(Clicked()), this, SLOT(slotResetDB()));
+
     connect(m_saveButton, SIGNAL(Clicked()), this, SLOT(slotSave()));
     connect(m_cancelButton, SIGNAL(Clicked()), this, SLOT(Close()));
 
@@ -87,6 +95,11 @@ bool GeneralSettings::Create()
                  "to the file and permissions must be set "
                  "accordingly. Features such as ID3 playcounts "
                  "and ratings depend on this being enabled."));
+    if (m_resetDBButton)
+        m_resetDBButton->SetHelpText(tr("This will clear all the MythMusic database tables allowing "
+                 "for a fresh start. NOTE: You may lose any manual or automatic changes made to "
+                 "a tracks metadata like rating or playcount unless you told MythMusic to "
+                 "write those to the tag."));
     m_cancelButton->SetHelpText(tr("Exit without saving settings"));
     m_saveButton->SetHelpText(tr("Save settings and Exit"));
 
@@ -97,6 +110,44 @@ bool GeneralSettings::Create()
     return true;
 }
 
+void GeneralSettings::slotResetDB(void)
+{
+    ShowOkPopup(tr("Are you sure you want to reset the music database?"),
+                this, SLOT(slotDoResetDB(bool)), true);
+}
+
+void GeneralSettings::slotDoResetDB(bool ok)
+{
+    if (ok)
+    {
+        gPlayer->stop(true);
+
+        MSqlQuery query(MSqlQuery::InitCon());
+
+        query.prepare("TRUNCATE music_albumart");
+        query.exec();
+        query.prepare("TRUNCATE music_albums");
+        query.exec();
+        query.prepare("TRUNCATE music_artists");
+        query.exec();
+        query.prepare("TRUNCATE music_directories");
+        query.exec();
+        query.prepare("TRUNCATE music_genres");
+        query.exec();
+        query.prepare("TRUNCATE music_playlists");
+        query.exec();
+        query.prepare("TRUNCATE music_songs");
+        query.exec();
+        query.prepare("TRUNCATE music_stats");
+        query.exec();
+
+        gMusicData->reloadMusic();
+
+        ShowOkPopup(tr("Music database has been cleared.\n"
+                       "You must now scan, rip or import some tracks."));
+    }
+}
+
 void GeneralSettings::slotSave(void)
 {
     gCoreContext->SaveSetting("CDDevice", m_musicCDDevice->GetText());
diff --git a/mythplugins/mythmusic/mythmusic/generalsettings.h b/mythplugins/mythmusic/mythmusic/generalsettings.h
index f931ff7..c1ee904 100644
--- a/mythplugins/mythmusic/mythmusic/generalsettings.h
+++ b/mythplugins/mythmusic/mythmusic/generalsettings.h
@@ -26,12 +26,15 @@ class GeneralSettings : public MythScreenType
     MythUITextEdit     *m_nonID3FileNameFormat;
     MythUICheckBox     *m_ignoreID3Tags;
     MythUICheckBox     *m_allowTagWriting;
+    MythUIButton       *m_resetDBButton;
     MythUIButton       *m_saveButton;
     MythUIButton       *m_cancelButton;
 
   private slots:
     void slotSave(void);
 
+    void slotResetDB(void);
+    void slotDoResetDB(bool ok);
 };
 
 #endif // GENERALSETTINGS_H
diff --git a/mythplugins/mythmusic/theme/default-wide/musicsettings-ui.xml b/mythplugins/mythmusic/theme/default-wide/musicsettings-ui.xml
index 71d0b15..a410861 100644
--- a/mythplugins/mythmusic/theme/default-wide/musicsettings-ui.xml
+++ b/mythplugins/mythmusic/theme/default-wide/musicsettings-ui.xml
@@ -65,6 +65,11 @@
             <position>650,425</position>
         </checkbox>
 
+        <button name="resetdatabase" from="basewidebutton">
+            <position>650,470</position>
+            <value>Reset music database</value>
+        </button>
+
         <textarea name="helptext" from="basetextarea">
             <area>100,540,1080,100</area>
             <cutdown>yes</cutdown>
diff --git a/mythplugins/mythmusic/theme/default/musicsettings-ui.xml b/mythplugins/mythmusic/theme/default/musicsettings-ui.xml
index a77dc9f..a6f1d83 100644
--- a/mythplugins/mythmusic/theme/default/musicsettings-ui.xml
+++ b/mythplugins/mythmusic/theme/default/musicsettings-ui.xml
@@ -59,12 +59,17 @@
         <textarea name="allowtagwriting_label" from="basetextarea">
             <area>20,360,380,40</area>
             <align>right,vcenter</align>
-            <value>Allow metadata to be written to tags:</value>
+            <value>Allow writing metadata to tags:</value>
         </textarea>
         <checkbox name="allowtagwriting" from="basecheckbox">
             <position>410,365</position>
         </checkbox>
 
+        <button name="resetdatabase" from="basewidebutton">
+            <position>410,410</position>
+            <value>Reset Database</value>
+        </button>
+
         <textarea name="helptext" from="basetextarea">
             <area>10,445,780,100</area>
             <cutdown>yes</cutdown>
-- 
1.7.10.2

